Ensure extension update notices only notify once within 24 hours, provide ability to disable all extension update notices

andyfeller opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
Fixes #9925

This PR brings the new extension update checking logic up to par with what is done for `gh` by allowing users to disable notices via `GH_NO_EXTENSION_UPDATE_NOTIFIER` environment variable while skipping them in various unattended scenarios. The result being `gh` will only check only it has been 24 hours since the executed extension was used.

Testing these changes is a bit tricky as there is a degree of complexity with the `update` package that is difficult to test again due to `ShouldCheckForExtensionUpdate()` logic looking at stdout and stdin.

In addition to the automated tests, I also ran these changes through acceptance tests:

```shell
$ set -o pipefail && GH_ACCEPTANCE_TOKEN=$(gh auth token --hostname github.com) GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -run ^TestExtensions$ github.com/cli/cli/v2/acceptance  | tparse --all go test 
┌──────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │           TEST           │             PACKAGE               │
│─────────┼─────────┼──────────────────────────┼───────────────────────────────────│
│  PASS   │   18.29 │ TestExtensions/extension │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestExtensions           │ github.com/cli/cli/v2/acceptance  │
└──────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 18.82s  │ github.com/cli/cli/v2/acceptance │  --   │  2   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

### Demo!

![Demo of GitHub CLI extension update enhancements](https://github.com/user-attachments/assets/07c60e6b-2080-4700-ba8e-e3971e3674d1)

<details>
<summary>
  <b>Made with <a href="https://github.com/charmbracelet/vhs/"><code>vhs</code></a></b>
</summary>
<p>

For more information, see https://github.com/charmbracelet/vhs/

```
Set Shell zsh
Output extension-updates.gif
Set Width 1200
Set Height 600

# Setup
Hide
Type "gh ext remove eco"
Enter
Type "rm -rf ~/.local/state/gh/extensions"
Enter
Type "clear"
Enter
Show

# Demo older non-local extension version triggering update notice
Type "gh notifies user of newer releases when dispatching extensions"
Sleep 1s
Ctrl+U 

Type "./bin/gh ext install jrnxf/gh-eco --pin v0.1.4"
Enter
Sleep 3s

Type "./bin/gh eco"
Enter
Sleep 1s

Type "andyfeller"
Enter
Sleep 2s
Escape
Sleep 5s

# Demo extension metadata captured
Ctrl+L 
Type "gh captures metadata about extension updates"
Sleep 1s
Ctrl+U 

Type "tree ~/.local/state/gh"
Sleep 500ms
Enter

Type "cat ~/.local/state/gh/extensions/gh-eco/state.yml"
Enter
Sleep 5s

# Demo repeated non-local extension does not trigger update notice
Ctrl+L 
Type "gh only notifies once within 24 hours per extension"
Sleep 1s
Ctrl+U 

Type "./bin/gh eco"
Enter
Sleep 1s

Type "andyfeller"
Enter
Sleep 2s
Escape
Sleep 5s

# Demo removing extension clears metadata
Ctrl+L 
Type "gh cleans up metadata when removing extensions"
Sleep 1s
Ctrl+U 

Type "tree ~/.local/state/gh"
Sleep 500ms
Enter

Type "./bin/gh ext remove eco"
Sleep 500ms
Enter

Type "tree ~/.local/state/gh"
Enter
Sleep 5s

# Demo notifications can be disabled via environment variable
Ctrl+L 
Type "users can disable this with environment variable"
Sleep 1s
Ctrl+U 

Type "./bin/gh ext install jrnxf/gh-eco --pin v0.1.4"
Enter
Sleep 3s

Type "GH_NO_UPDATE_NOTIFIER=1 ./bin/gh eco"
Enter
Sleep 1s

Type "andyfeller"
Enter
Sleep 1s
Escape

Type "tree ~/.local/state/gh"
Enter
Sleep 5s

# Demo how metadata is removed when installing extensions
Ctrl+L 
Type "gh cleans up metadata when installing extensions"
Sleep 1s
Ctrl+U 

Type "./bin/gh eco"
Enter
Sleep 1s

Type "andyfeller"
Enter
Sleep 1s
Escape

Type "tree ~/.local/state/gh"
Sleep 500ms
Enter

Type "rm -rf ~/.local/share/gh/extensions/gh-eco"
Sleep 500ms
Enter

Type "./bin/gh ext install jrnxf/gh-eco --pin v0.1.4"
Sleep 500ms
Enter

Type "tree ~/.local/state/gh"
Sleep 500ms
Enter
Sleep 5s
```

</p>
</details>

andyfeller committed:
Bring extension update check in line with gh check

This commit is a bit of refactoring to bring the extension update checking logic up to par with what is done with `gh` including creation of state file per extension and listening to env vars for disabling version checking.

This work is not complete as it does not address necessary test changes.

andyfeller committed:
Refactor extension command testing

This commit allows for stubbing/mocking the function used for checking update release info.  Additionally, this does a minor refactor moving from non-blocking to blocking channel.

BagToad commented on internal/update/update.go:
```diff
@@ -31,6 +33,49 @@ type StateEntry struct {
+
+func CheckForExtensionUpdate(em extensions.ExtensionManager, ext extensions.Extension, stateFilePath string) (*ReleaseInfo, error) {
+	stateEntry, _ := getStateEntry(stateFilePath)
+	if stateEntry != nil && time.Since(stateEntry.CheckedForUpdateAt).Hours() < 24 {
+		return nil, nil
+	}
```
_I don't think this feedback is blocking this PR..._

----

Do you think this interval could get annoying for pinned extension users?

_Disclaimer: I say this as someone who does not leverage the extension pinning feature, I could be misunderstanding something_. 

My guess is that someone pins an extension because they do not want whatever the latest version is. I recognize that this user may still want to know when a new version becomes available, but do we want to continue notifying them at the same 24h interval as other extensions? Would a longer interval make more sense for these users that have already expressed that they do not want to use the latest version?

If we leave pinned extensions as checking and notifying once every 24 hours, I foresee a situation where a user is notified of an available update at most once a day for an extension that they have no intention of updating. In that case, we are giving the user almost no choice but to disable the extension update check altogether if they feel disrupted by the update notification.

Direct Qs:
- Am I correct in thinking that this might be a problem for some folks, and do I understand the pinning feature correctly?
- What do you think about adjusting the interval for pinned extensions?
- Is there something else you could think of to allow a pinned extension user to avoid receiving such frequent notifications, while leaving the updating notification feature on?

BagToad replied:
Perhaps this is a bit spicy, but what if we disable the update checking logic for pinned extensions altogether? 

I can loosely see the logic behind a user expressing that they want `x` version means we shouldn't tell them about available updates. And, as long as that behavior is documented, I can see this making sense to do.

andyfeller replied:
Let me think on this question a bit, but here's an alternative view: **users pin a version to ensure a specific SHA is installed.**

Like GitHub Actions, the version is:

- tag / release
- branch
- SHA

tag/release and branch are moving targets, not idempotent but mutable.  a SHA is a distinct identifier so more reliable.

Pining an extension is more than casually preventing someone from upgrading the extension &mdash; which can be done in a few ways with `gh ext` commands &mdash; so I'm not 💯 this is meant to avoid upgrading. 🤔

That said, let me re-read the code here versus the `gh` code.  I think in the `gh` code it is actually checking that the version looked up differs from the one that's recorded in the state file, which means users are only notified of new versions.  If that's the case, then we should make sure the extension release checking logic is going the same, so users are only notified when a new version is seen for the first time.

Commit:
Initial extension update check testing

First pass at implementing basic test around extension state checking behavior, wanting to discus with team about level of testing to perform and whether this is really the right place.

andyfeller commented:
After discussing the state of this PR with @BagToad last week, I've been working on the tests around some of the changes as I'd like to see more of the extension update logic tested.  I opened https://github.com/cli/cli/pull/9985 for discussion because the original logic is not really tested and I'm unsure about the level of value in testing at a particular level or not.

Commit:
Enhance extension manager and tests

- extension manager has been made responsible for extension update directory information
- extension manager has been enhanced to ensure that previously left extension update metadata is deleted before installing an extension
- extension manager has been enhanced to ensure that extension update metadata is deleted when extension is being removed
- refactored extension command tests for manager change, ensuring previous and expected states along with returned release info
- refactored extension manager tests for ensuring previous extension update entries are removed before installing extension
- created extension manager test for installing local extension
- centralized logic for checking and ensuring extension name is "gh-" prefixed

Commit:
Remove Extension.FullName() from interface

- this was something I did in the original implementation of this improved extension update notification logic
- discovering other parts of the extension manager code that were working with non-prefixed extension names motivated a different approach
- the extension manager code that requires the extension be prefixed has been enhanced to use the centralized ensurePrefixed() logic, making the need for this on the extension unnecessary

andyfeller committed:
Test whether extension update should be checked

andyfeller committed:
Expand update tests for local extensions

andyfeller committed:
Remove unnecessary Extension.FullName()

andyfeller committed:
Refactor local installation manager test by OS

- Local extensions on Windows are not symlinks but files whose contents is the path to the local extension directory
- Local extensions elsewhere are symlinks

andyfeller committed:
Add build constraint for non-Windows systems

andyfeller committed:
Add build constraint for non-Windows systems

BagToad commented on internal/update/update.go:
```diff
@@ -31,6 +33,54 @@ type StateEntry struct {
+	if os.Getenv("CODESPACES") != "" {
+		return false
+	}
+	return updaterEnabled != "" && !IsCI() && IsTerminal(os.Stdout) && IsTerminal(os.Stderr)
```
Can we expand on why `updaterEnabled` is here and not in the extension updater logic?

andyfeller replied:
I agree this is a little too obtuse as-is.

In retrospect after our paired review, I got to thinking whether the duplicate logic between the 2 functions was valuable.  I began thinking this specific criteria should be moved out of to the gh update logic in `cmd.go` so it could be reused:

```go
func checkForUpdate(ctx context.Context, f *cmdutil.Factory, currentVersion string) (*update.ReleaseInfo, error) {
	if updaterEnabled == "" || !update.ShouldCheckForUpdate() {
		return nil, nil
	}
	httpClient, err := f.HttpClient()
	if err != nil {
		return nil, err
	}
	repo := updaterEnabled
	stateFilePath := filepath.Join(config.StateDir(), "state.yml")
	return update.CheckForUpdate(ctx, httpClient, stateFilePath, repo, currentVersion)
}
```

BagToad commented on pkg/cmd/root/extension.go:
```diff
@@ -7,40 +7,36 @@ import (
-						Pinned:         ext.IsPinned(),
-						URL:            ext.URL(),
-					}
+				rel, err := checkFunc(em, ext)
```
Could we rename `checkFunc` to something that relates to `ReleaseInfo` so that it's more readily understandable what this is doing and what `rel` is?

BagToad commented on internal/update/update.go:
```diff
@@ -31,6 +33,54 @@ type StateEntry struct {
+}
+
+func CheckForExtensionUpdate(em extensions.ExtensionManager, ext extensions.Extension, now time.Time) (*ReleaseInfo, error) {
+	if ext.IsLocal() {
```
@andyfeller can you leave a comment in the code about why this is here?

BagToad commented on pkg/cmd/extension/manager.go:
```diff
@@ -46,7 +47,10 @@ type Manager struct {
 	return &Manager{
-		dataDir:    config.DataDir,
+		dataDir: config.DataDir,
+		updateDir: func() string {
```
Maybe there is a better name for this internal func to differentiate between this and the exported one 🤔

andyfeller replied:
Going to think on this one a bit because I'm still not sure of something better 🤔

andyfeller replied:
honestly, I couldn't really think of a better name outside of calling it `stateDir` which is what `config` uses for related directory.

unless there is a strong recommendation, I don't know if this is blocking nor does it change that `dataDir` is related to `config.DataDir` 🤷

BagToad commented on pkg/cmd/extension/manager.go:
```diff
@@ -802,3 +825,23 @@ func codesignBinary(binPath string) error {
+}
+
+// ensurePrefixed just makes sure that the provided extension name is prefixed with "gh-".
+func ensurePrefixed(name string) string {
```
@andyfeller says maybe we should change the name to something like `normalizeExtension`

BagToad commented on pkg/cmd/extension/manager.go:
```diff
@@ -802,3 +825,23 @@ func codesignBinary(binPath string) error {
+// cleanExtensionUpdateDir should be used before installing extensions to avoid past metadata creating problems.
+func (m *Manager) cleanExtensionUpdateDir(name string) error {
+	updatePath := m.UpdateDir(name)
+	if _, err := os.Stat(updatePath); err == nil {
```
Consider checking for error type being a file that does not exist specifically. 

Example:

https://github.com/cli/cli/blob/b3f74900f2f078e5fa52483775ed316595a589a3/pkg/cmd/extension/manager.go#L214-L226

BagToad commented on pkg/cmd/extension/manager.go:
```diff
@@ -802,3 +825,23 @@ func codesignBinary(binPath string) error {
 	return cmd.Run()
 }
+
+// cleanExtensionUpdateDir should be used before installing extensions to avoid past metadata creating problems.
+func (m *Manager) cleanExtensionUpdateDir(name string) error {
+	updatePath := m.UpdateDir(name)
+	if _, err := os.Stat(updatePath); err == nil {
+		if err := os.RemoveAll(updatePath); err != nil {
+			return fmt.Errorf("failed to remove previous extension update state: %w", err)
+		}
+	}
+
+	return nil
+}
```
Might need to consider bubbling up the errors from `os.stat` due to something other than the file not existing.

andyfeller replied:
Just pushed changes, don't like the aesthetics of the change but they're done.

BagToad commented on pkg/cmd/extension/manager_other_test.go:
```diff
@@ -0,0 +1,36 @@
+//go:build !windows
```
From @andyfeller @BagToad conversation:

- Need to merge in changes from https://github.com/cli/cli/pull/9933
- Make sure you are cleaning up the extension update directory

andyfeller replied:
In pulling these changes into the branch, I made a minor change for consistency with other PR changes. 👍

BagToad commented on pkg/cmd/root/extension_test.go:
```diff
@@ -16,6 +17,7 @@ func TestNewCmdExtension_Updates(t *testing.T) {
 	tests := []struct {
 		name               string
 		extCurrentVersion  string
+		extFullName        string
```
@andyfeller - this should be removed because this is not being used by the tests.

BagToad reviewed (changes requested):
Discussed with @andyfeller

andyfeller committed:
Implement partial feedback before trunk update

andyfeller committed:
Merge branch 'trunk' into andyfeller/cli-cli-8183-timeout-pt2

andyfeller committed:
Implement remaining PR feedback

After merging in upstream changes regarding local extensions, this commit addresses remaining PR feedback while also bringing the newly merged tests into alignment with other changes.

andyfeller committed:
Fix failig local extension tests

Extensions have to be gh- prefixed and these newer tests were failing because of expectation.

andyfeller commented on internal/ghcmd/cmd.go:
```diff
@@ -216,29 +219,8 @@ func printError(out io.Writer, err error, cmd *cobra.Command, debug bool) {
-
 func checkForUpdate(ctx context.Context, f *cmdutil.Factory, currentVersion string) (*update.ReleaseInfo, error) {
-	if !shouldCheckForUpdate() {
+	if updaterEnabled == "" || !update.ShouldCheckForUpdate() {
```
In expanding the `update` package to also handle extensions, it felt like the right decision not to duplicate the logic so we had slightly different logic for `gh` vs extension should check logic.  The only difference was passing in the statically linked variable here.  By moving this unique criteria here, it makes `ShouldCheckForUpdate` reusable.

BagToad replied:
I like it!

andyfeller commented on pkg/cmd/extension/manager_test.go:
```diff
@@ -720,26 +752,31 @@ func TestManager_UpgradeExtension_GitExtension_Pinned(t *testing.T) {
-	m := newTestManager(extManagerDir, nil, nil, ios)
-	fakeExtensionName := "local-ext"
+	m := newTestManager(dataDir, updateDir, nil, nil, ios)
+	fakeExtensionName := "gh-local-ext"
```
`local-ext` created a problem without being prefixed as extension manager works to ensure all extension directories are `gh-` prefixed for both the data and update directories.

BagToad replied:
Ah yes! Makes sense

andyfeller commented on pkg/cmd/extension/manager_test.go:
```diff
@@ -720,26 +752,31 @@ func TestManager_UpgradeExtension_GitExtension_Pinned(t *testing.T) {
 }
 
 func TestManager_Install_local(t *testing.T) {
-	extManagerDir := t.TempDir()
```
Renamed this to be inline with all of the other tests but also because it is called `dataDir` within the extension manager, which now has 2 directories to care about.

andyfeller commented on pkg/cmd/root/extension.go:
```diff
@@ -7,40 +7,36 @@ import (
-
-func NewCmdExtension(io *iostreams.IOStreams, em extensions.ExtensionManager, ext extensions.Extension) *cobra.Command {
-	updateMessageChan := make(chan *extensionReleaseInfo)
+func NewCmdExtension(io *iostreams.IOStreams, em extensions.ExtensionManager, ext extensions.Extension, checkExtensionReleaseInfo func(extensions.ExtensionManager, extensions.Extension) (*update.ReleaseInfo, error)) *cobra.Command {
```
Open to ideas on alternatives to `checkExtensionReleaseInfo` function approach to make this testable. 🤷

williammartin replied:
Seems fine to me.

BagToad commented on internal/update/update.go:
```diff
@@ -31,6 +33,45 @@ type StateEntry struct {
 	LatestRelease      ReleaseInfo `yaml:"latest_release"`
 }
 
+func ShouldCheckForUpdate() bool {
```
@andyfeller Would you be okay adding a comment to this and other new exported functions?

andyfeller committed:
Comment exported functions

williammartin commented on pkg/cmd/extension/manager.go:
```diff
@@ -829,3 +852,26 @@ func codesignBinary(binPath string) error {
 	return cmd.Run()
 }
+
+// cleanExtensionUpdateDir deletes extension state directory to avoid problems reinstalling extensions.
```
@andyfeller can you list specific problems that you have in mind here? I can't think of any so I'm not sure I'm missing something.

andyfeller replied:
Absolutely!

Normally, we expect users are removing extensions using `gh ext remove <extension>`, however I've heard people mention they or others can just `rm -rf ~/.local/share/gh/extensions` or remove the specific extension subdirectory.

If a user did this but did not remove the extension update state information in `~/.local/state/gh/extensions/<extension>`, then extension manager can get into a weird state given extensions are not globally unique.

For example,

1. `gh ext install andyfeller/gh-ext`
2. `rm -rf ~/.local/share/gh/extensions/gh-ext`
3. `gh ext install williammartin/gh-ext`

Are these forks?  Just coincidentally named extensions?  What should the extension manager do if the old state was left around?

Instead, `cleanExtensionUpdateDir` is intended to clean up any lingering information to ensure installing an extension has a clean slate.

williammartin replied:
Thanks for the explanation. Does this have any actual practical implications right now?

I mean, if it's been less than 24 hours, we don't show the update notice. If it has been more than 24 hours we fetch the information for `williammartin/gh-ext`?

andyfeller replied:
PS: `cleanExtensionUpdateDir` is used by the manager when installing or removing an extension.

**Removing an extension**

https://github.com/cli/cli/blob/6bbd70353b03975bc230759fadf9d58a418e226d/pkg/cmd/extension/manager.go#L571-L584

**Installing a local extension**

https://github.com/cli/cli/blob/6bbd70353b03975bc230759fadf9d58a418e226d/pkg/cmd/extension/manager.go#L207-L212

**Installing a binary extension**

https://github.com/cli/cli/blob/6bbd70353b03975bc230759fadf9d58a418e226d/pkg/cmd/extension/manager.go#L334-L343

**Installing a script extension**

https://github.com/cli/cli/blob/6bbd70353b03975bc230759fadf9d58a418e226d/pkg/cmd/extension/manager.go#L410-L428

williammartin replied:
Nudge on my previous message because you might have missed it, commenting right after me.

andyfeller replied:
> Nudge on my previous message because you might have missed it, commenting right after me.

Definitely didn't see your horse cross the line faster than the web UI refreshed to show while I was following up.  Thanks for the nudge.

> Thanks for the explanation. Does this have any actual practical implications right now?
> 
> I mean, if it's been less than 24 hours, we don't show the update notice. If it has been more than 24 hours we fetch the information for `williammartin/gh-ext`?

Currently, the `update` package is only using the last checked information for determining whether it is too soon to check for an update.

By leaving this information around after removing an extension, it will either 1) lead to an immediate check because the last check time in the file is quite a bit older or 2) manual modifications to this file setting this date into the future would prevent notifications from going out. 

That said, I also don't know what the future holds around extension consumption, however I know there are ideas like #6118 that might depend upon information like this to be cleaned up.  For that reason, I think cleaning it up is important.

Lastly, I thought this was a scenario you yourself mentioned when I first begun on this work.

---

With all of that said, @williammartin, is there a concern motivating this question?

williammartin replied:
> By leaving this information around after removing an extension, it will either 1) lead to an immediate check because the last check time in the file is quite a bit older or 2) manual modifications to this file setting this date into the future would prevent notifications from going out.

This doesn't seem any different than installation of an extension for the first time, upon which we are going to do a check for an update immediately (which by the way, might be a bit annoying if I specifically pinned an extension on install, but isn't the end of the world).

What am I missing?

> Lastly, I thought this was a scenario you yourself mentioned when I first begun on this work.

Right, I mentioned that it would be nice if they could live in the extension installation location so that they got cleaned up automatically (I understand why that wasn't such an easy solution). However, then, and now, I couldn't point to their being a specific issue with leaving them around.

> is there a concern motivating this question?

Yeh, three:
 1. I'm concerned I don't understand if there is a change in behaviour that would occur by leaving these around.
 2. I'm concerned that I, or someone else will have the same question, only less context in the future, so I'd like any specific changes in behaviour are captured in a comment for the future rather than non-specific "problems". If there are no specific changes in behaviour, and we choose to keep it, I want that to be expressed in the comment.
 3. I'm generally just a bit concerned about the maintainability of a solution that requires `cleanExtensionUpdateDir` to be placed in three different locations in the three install functions, and in ways that aren't necessarily enforced by tests. The extra cognitive load understanding how the positioning of these calls relates to the code around them creates some friction (e.g. does it functionally matter if the call is inside the `dryRun` check block? Probably not, but I have to reason that out by understanding dryRun is currently only relevant for upgrades).

All-in-all, I think it's generally hard to reason about state management, and it only gets harder over time. I'm not really sure how to better express 3, other than it gives me the heebie-jeebies based on some reason I'm not skilled enough to articulate. Heebie-jeebies is non-blocking but it's enough to make me want 1 and 2 addressed.

andyfeller replied:
Those are all completely legitimate concerns! 💯

---

> > Lastly, I thought this was a scenario you yourself mentioned when I first begun on this work.
> 
> Right, I mentioned that it would be nice if they could live in the extension installation location so that they got cleaned up automatically (I understand why that wasn't such an easy solution)

😞 Definitely a nuanced and difficult situation for several reasons:

1. Script based extensions are locally cloned git repositories, so adding this metadata there introduces some unknown behaviors
1. Local extensions don't support current/latest versions and work differently on Windows _(file containing a path to the local extension)_ vs others _(symlink to directory)_
1. Hosting this state in a place where an extension author could collide or override it seems generally bad
1. Aligning this to [XDG standard](https://specifications.freedesktop.org/basedir-spec/latest/) feels like this is state information that should be outside of the extension code

---

> > is there a concern motivating this question?
> 
> Yeh, three:
> 
> 1. I'm concerned I don't understand if there is a change in behaviour that would occur by leaving these around.
> 2. I'm concerned that I, or someone else will have the same question, only less context in the future, so I'd like any specific changes in behaviour are captured in a comment for the future rather than non-specific "problems". If there are no specific changes in behaviour, and we choose to keep it, I want that to be expressed in the comment.
> 3. I'm generally just a bit concerned about the maintainability of a solution that requires `cleanExtensionUpdateDir` to be placed in three different locations in the three install functions, and in ways that aren't necessarily enforced by tests. The extra cognitive load understanding how the positioning of these calls relates to the code around them creates some friction (e.g. does it functionally matter if the call is inside the `dryRun` check block? Probably not, but I have to reason that out by understanding dryRun is currently only relevant for upgrades).
> 
> All-in-all, I think it's generally hard to reason about state management, and it only gets harder over time. I'm not really sure how to better express 3, other than it gives me the heebie-jeebies based on some reason I'm not skilled enough to articulate. Heebie-jeebies is non-blocking but it's enough to make me want 1 and 2 addressed.

Again, all completely valid points.

1. What else can we do to ensure we're confident about this given the efforts to consolidate related code into `update` package and incorporate extension update state in  `manager_test.go` cases?

   _I'm genuinely asking as I want to ensure we do know and have confidence._ 🙇 

1. I can abide 🤜 🤛 

1. Is this a limitation of the current extension architecture, current extension tests suite, or simply additional tests outside of the enhancements made to the install tests so far?

I can certainly :nod: that we should spend some time re-evaluating the extension architecture _(manager, kind-specific types, manifests, etc)_ especially with an eye for testability and maintainability.

So I'd appreciate anything I can do to ensure our confidence.

andyfeller replied:
> > By leaving this information around after removing an extension, it will either 1) lead to an immediate check because the last check time in the file is quite a bit older or 2) manual modifications to this file setting this date into the future would prevent notifications from going out.
> 
> This doesn't seem any different than installation of an extension for the first time, upon which we are going to do a check for an update immediately (which by the way, might be a bit annoying if I specifically pinned an extension on install, but isn't the end of the world).
> 
> What am I missing?

You're absolutely correct; this would be no different than if we install an extension for the first time. 👍  So, it isn't a problem.

The problem/concern is #2 where an older metadata file where the last checked date was manually set into the future that prevents the user from getting a notice.  That said, as we're discussing this together at the time of writing, this is something we have no control over a user or a nefarious user doing intentionally.

andyfeller replied:
Having discussed this further with @williammartin, I'm going to expand the documentation of this function to capture the expectations discussed here for posterity.

andyfeller replied:
@williammartin : will defer to you for resolving this conversation given cli/cli@0a1c9be656997623b08b424e8ebb528085dfe733 as I want to ensure you are satisfied 🙇

williammartin replied:
LGTM. Though I'd maybe say "the only effect" -> "the only known effect"

andyfeller commented on internal/ghcmd/cmd.go:
```diff
@@ -25,11 +25,14 @@ import (
+// used to control whether users are notified of newer GitHub CLI releases.
+// Development builds leave this empty as impossible to determine if newer or not.
+// For more information, <https://github.com/Homebrew/homebrew-core/blob/master/Formula/g/gh.rb>.
 var updaterEnabled = ""
```
Update the comment here calling out that this has to be set to `cli/cli` in order to check the right repo for newer releases.

andyfeller commented on internal/update/update.go:
```diff
@@ -31,7 +33,48 @@ type StateEntry struct {
-// CheckForUpdate checks whether this software has had a newer release on GitHub
+// ShouldCheckForUpdate decides whether we check for updates for the GitHub CLI and/or GitHub CLI extensions based on user preferences and current execution context.
+func ShouldCheckForUpdate() bool {
+	if os.Getenv("GH_NO_UPDATE_NOTIFIER") != "" {
```
Talking with @williammartin, we recognize that some users will not want update notices about extensions despite our concerns with security or desires to support extension authors with newer releases.  **To that end, we will use a separate environment variable to allow those users to disable all extension update notices for now until we have a more robust mechanism to control extension-specific notices.**

To that end, we'll duplicate this function as its likely that the logic will continue to evolve separate for core `gh` versus extensions 👍

andyfeller commented on pkg/cmd/extension/manager.go:
```diff
@@ -829,3 +852,26 @@ func codesignBinary(binPath string) error {
+// cleanExtensionUpdateDir deletes extension state directory to avoid problems reinstalling extensions.
+func (m *Manager) cleanExtensionUpdateDir(name string) error {
+	updatePath := m.UpdateDir(name)
+	if _, err := os.Stat(updatePath); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to check previous extension update state: %w", err)
+	}
+	if err := os.RemoveAll(updatePath); err != nil {
```
@williammartin mentioned that maybe this block isn't actually needed as `os.RemoveAll` can handle the directory being there or not

williammartin committed:
Collapse dryrun checks in ext bin upgrade

williammartin commented:
This is non blocking because I really want to get this PR shipped and it all relates to code structure.

As I looked at this PR again, I was interested in the "ownership" of the `UpdateDir`. Ostensibly, the `ExtensionManager` owns it, returning the filepath to it from the `UpdateDir` method. It also cleans it up on `install` and `remove`. However, it doesn't write to it, instead letting others do that (i.e. the `update.go` functions).

Therefore the contract is:
 * `ExtensionManager` owns the lifecycle of the dir
 * `ExtensionManager` allows for others to interact with the contents of the dir

This is alright, but in practice, the only collaborator for the second point is the `update.go` functions. It might make more sense in a follow up to bring the update logic into the `ExtensionManager` itself, so that the cobra lifecycle can call something like `em.CheckForUpdate(ext)`.

I understand that currently the desire is to have parallels between the `gh` updating, and the `ext` manager updating logic. I believe that we could abstract a common module to be shared by these two updating-aware features, and have it called from `ExtensionManager`.

andyfeller committed:
Capture greater detail on updaterEnabled

This commit expands the in-line docs around updaterEnabled package variable used to affect release checking.

Along with clarifying specific details discovered when talking with @williammartin, I'm also removing a useless local variable.

andyfeller committed:
Separate logic for checking updates

During discussion in cli/cli#9934, we can to the conclusion that the logic around checking for core GitHub CLI updates would diverge from GitHub CLI
extension updates over time.  To that end, this commit splits that logic into a separate function with a new environment variable.

andyfeller committed:
Simplifying cleanExtensionUpdateDir logic

This function was doing some unnecessary heavy lifting detecting if the directory being deleted actually existed when `os.RemoveAll()` would handle directories that exist or not.

andyfeller committed:
Expand docs on cleaning extension update dir

This summarizes a conversation within cli/cli#9934 PR regarding how the new extension update directory metadata would affect gh if not cleaned up.

BagToad reviewed (approved):
Tentatively approving this PR but @andyfeller and I want to give time for @williammartin to respond to any remaining comments before merging 🙂

Edit: thank you @andyfeller for the title edit ✨

williammartin commented on pkg/cmd/root/help_topic.go:
```diff
@@ -86,8 +86,12 @@ var HelpTopics = []helpTopic{
+			By default, gh checks for new releases once every 24 hours and displays an upgrade notice on standard
+			error if a newer version was found.
+
+			%[1]sGH_NO_EXTENSION_UPDATE_NOTIFIER%[1]s: set to any value to disable GitHub CLI extension update notifications.
```
Maybe we should indicate here that the check only occurs when the extension is actually called?

williammartin replied:
There's no file changed here for me to leave a comment on, but I think I'd also expect to see something relating to extension update logic, or pointing to this env var in the `gh ext --help`

williammartin replied:
Also no file changed here for me to comment on but this currently looks like:

```
A new release of sonar is available: 81a4ce86e027f31d306883c25a71b5d05b007e2e → 9977d8e535f98d57b15f0b5cb1dc887205acd2e4
To upgrade, run: gh extension upgrade sonar
https://github.com/andyfeller/gh-sonar.git
```

"release" is maybe not the most accurate term for this.

It does make me think that the current upgrade approach around `GitKind` extensions is probably not very friendly, with a potential notification every 24 hours for any new git commits, and no obvious way to understand whether the git commit is interesting. Non-blocking.

andyfeller replied:
It isn't something we needed to mention in the core GH update notifier env var above, but we can mention it in both.

williammartin replied:
Sorry you lost me on this. What are you saying we can mention in both? I was just expecting some additional clarity on top of:

```
By default, gh checks for new releases once every 24 hours and displays an upgrade notice on standard error if a newer version was found.
```

This is only part of the story, it's missing the fact this check only occurs when the extension is run, for that specific extension.

williammartin commented on pkg/cmd/root/help_topic.go:
```diff
@@ -86,8 +86,12 @@ var HelpTopics = []helpTopic{
+			By default, gh checks for new releases once every 24 hours and displays an upgrade notice on standard
+			error if a newer version was found.
+
+			%[1]sGH_NO_EXTENSION_UPDATE_NOTIFIER%[1]s: set to any value to disable GitHub CLI extension update notifications.
```
I think I'd expect to see something in the `gh ext --help` indicating extension upgrade behaviour, and maybe pointing to this env var.

williammartin replied:
Also no file changed here for me to comment on but this currently looks like:

```
A new release of sonar is available: 81a4ce86e027f31d306883c25a71b5d05b007e2e → 9977d8e535f98d57b15f0b5cb1dc887205acd2e4
To upgrade, run: gh extension upgrade sonar
https://github.com/andyfeller/gh-sonar.git
```

"release" is maybe not the most accurate term for this.

It does make me think that the current upgrade approach around `GitKind` extensions is probably not very friendly, with a potential notification every 24 hours for any new git commits, and no obvious way to understand whether the git commit is interesting. Non-blocking.

williammartin reviewed (approved):
I have a few non-blocking notes.

This makes a **significant** difference to the latency when I'm running `gh ext --help`, thanks.

```
➜  cli git:(andyfeller/cli-cli-8183-timeout-pt2) hyperfine -w2 "gh sarif --help" "/Users/williammartin/workspace/cli/bin/gh sarif --help" "GH_NO_EXTENSION_UPDATE_NOTIFIER=true /Users/williammartin/workspace/cli/bin/gh sarif --help"
Benchmark 1: gh sarif --help
  Time (mean ± σ):      1.087 s ±  0.006 s    [User: 0.078 s, System: 0.026 s]
  Range (min … max):    1.073 s …  1.092 s    10 runs

Benchmark 2: /Users/williammartin/workspace/cli/bin/gh sarif --help
  Time (mean ± σ):      62.4 ms ±   4.7 ms    [User: 54.0 ms, System: 10.6 ms]
  Range (min … max):    58.6 ms …  76.4 ms    40 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 3: GH_NO_EXTENSION_UPDATE_NOTIFIER=true /Users/williammartin/workspace/cli/bin/gh sarif --help
  Time (mean ± σ):      62.6 ms ±   5.6 ms    [User: 54.4 ms, System: 10.6 ms]
  Range (min … max):    58.5 ms …  92.2 ms    47 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  /Users/williammartin/workspace/cli/bin/gh sarif --help ran
    1.00 ± 0.12 times faster than GH_NO_EXTENSION_UPDATE_NOTIFIER=true /Users/williammartin/workspace/cli/bin/gh sarif --help
   17.41 ± 1.30 times faster than gh sarif --help
```

andyfeller committed:
Enhance help docs on ext upgrade notices

- update `gh help environment` to include that upgrade notices only happen when gh or extensions are executed
- update `gh ext --help` to include reference to upgrade notices and points to `gh help environment` for info on disabling
